### PR TITLE
feat: add `try_to_number` function

### PR DESCRIPTION
**Pull request by rafafrdz**
Following the implementation of `to_number` (merged in #629 ).

This PR introduces the implementation of the `try_to_number` functionality, closely replicating Spark's `try_to_number` capabilities. The function enables safe conversion of string columns into numeric types by parsing formatted strings according to locale-aware number patterns, returning a decimal representation or null on invalid input.

Spark Ref: [`try_to_number` doc](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.try_to_number.html)
Databricks Ref: [`try_to_number` doc](https://docs.databricks.com/aws/en/sql/language-manual/functions/try_to_number)

In this PR, we:

- [x] Add support for the `try_to_number` Spark SQL function  
- [x] Add doc tests
- [x] Fix grouping comma condition in `to_number`
- [x] Fix the angled brackets in the format string (they must be both or none)

part of #508

### CODE REVIEW

**Comment by rafafrdz** [comment]
In this PR, I extracted the logic that handles Array values from to_number to avoid code duplication. This allows me to reuse the SparkToNumber behavior by simply wrapping the value handler function to return Null whenever SparkToNumber throws an exception.

File: crates/sail-plan/src/extension/function/string/spark_try_to_number.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+impl Default for SparkTryToNumber {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ScalarUDFImpl for SparkTryToNumber {
+    fn as_any(&self) -> &dyn Any {
+        self.inner.as_any()
+    }
```
Comment (lonless9) [inline crates/sail-plan/src/extension/function/string/spark_try_to_number.rs]: Could we return `self` instead?

File: crates/sail-plan/src/extension/function/string/spark_to_number.rs [hunk 2]
```diff
@@ -98,49 +99,72 @@ impl ScalarUDFImpl for SparkToNumber {
...
-            SparkToNumber::NAME,
-            args.len()
-        );
+pub trait SparkToNumberInnerImpl {
+    fn value_to_number(
+        value: Option<&str>,
+        number_components: &NumberComponents,
+        value_regex: &Regex,
+        format_spec: &RegexSpec,
+    ) -> Result<ScalarValue>;
```
Comment (lonless9) [inline crates/sail-plan/src/extension/function/string/spark_to_number.rs]: Maybe we can follow `SparkTryToBinary` instead?

File: python/pysail/tests/spark/function/test_try_to_number.txt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
++----------+------------+
+|-123456.32|decimal(8,2)|
++----------+------------+
+
+>>> spark.sql("SELECT try_to_number('-$123,456.32-', 'MI$999,999.99S') AS result, typeof(try_to_number('-$123,456.32-', 'MI$999,999.99S')) AS type").show(truncate=False)
++---------+------------+
+|result   |type        |
++---------+------------+
+|123456.32|decimal(8,2)|
++---------+------------+
```
Comment (lonless9) [inline python/pysail/tests/spark/function/test_try_to_number.txt:69]: I'm curious whether Spark would return a positive or negative number in this case?

File: crates/sail-plan/src/extension/function/string/spark_to_number.rs [hunk 4]
```diff
@@ -510,18 +534,27 @@ fn match_grouping(value_captures: &Captures, format_spec: &RegexSpec) -> Result<
     let format_positions = get_grouping_positions(&format_numbers);
     let number_positions = get_grouping_positions(&numbers);
 
+    // Check if the positions of the number groupings are in the format groupings
+    let grouping_cond = number_positions.is_subset(&format_positions);
+
```
Comment (lonless9) [inline crates/sail-plan/src/extension/function/string/spark_to_number.rs]: What will format `9,999` with value `1000` look like?

**Review by lonless9 — COMMENTED** [review]
This looks great! I've left some comments.

File: crates/sail-plan/src/extension/function/string/spark_to_number.rs (hunk 2 above)
Comment (rafafrdz) [reply crates/sail-plan/src/extension/function/string/spark_to_number.rs]: Got it!

File: crates/sail-plan/src/extension/function/string/spark_to_number.rs (hunk 4 above)
Comment (rafafrdz) [reply crates/sail-plan/src/extension/function/string/spark_to_number.rs]: Yeah, you're right! The code below does check if it's a subset, but also ensures that the grouping separator characters are at the expected positions, not just that the separators exist. As you said, just checking for a subset isn't enough. Thanks!

```rust
    for (pos, sep) in &format_positions {
        if pos < &numbers.len() && numbers.chars().rev().nth(*pos) != Some(*sep) {
            return exec_err!("Malformed integer format related groupings: {format_numbers}.");
        }
    }
```

I'll delete it then!

File: python/pysail/tests/spark/function/test_try_to_number.txt (hunk 3 above)
Comment (rafafrdz) [reply python/pysail/tests/spark/function/test_try_to_number.txt:69]: The way `to_number` and `try_to_number` work under the hood is kind of black magic, you think you understand it until you run into some weird formatting edge case 😅

It's not really obvious how they thought /implemented their behaviour... and the documentation as well...

File: crates/sail-plan/src/extension/function/string/spark_try_to_number.rs (hunk 1 above)
Comment (rafafrdz) [reply crates/sail-plan/src/extension/function/string/spark_try_to_number.rs]: done!

File: crates/sail-plan/src/extension/function/string/spark_to_number.rs (hunk 2 above)
Comment (rafafrdz) [reply crates/sail-plan/src/extension/function/string/spark_to_number.rs]: done! I re-did it using that approach 😄

**Comment by rafafrdz** [comment]
Hey @lonless9 do you know anything about the clippy error that the GitHub action threw? it has nothing to do with `try_to_number`. Besides, I ran in my local `cargo clippy --all-targets --all-features -- -D warnings` and nothing happened 😅

**Comment by lonless9** [comment]
> Hey @lonless9 do you know anything about the clippy error that the GitHub action threw? it has nothing to do with `try_to_number`. Besides, I ran in my local `cargo clippy --all-targets --all-features -- -D warnings` and nothing happened 😅

You may need to update your rust toolchain version via `rustup update`.

**Comment by rafafrdz** [comment]
> > Hey @lonless9 do you know anything about the clippy error that the GitHub action threw? it has nothing to do with `try_to_number`. Besides, I ran in my local `cargo clippy --all-targets --all-features -- -D warnings` and nothing happened 😅
> 
> You may need to update your rust toolchain version via `rustup update`.

Sorry, I didn't explain myself well 😅 What I wanted to say is that there is an error when the GitHub action runs clippy that which has nothing to do with any of the code I did.

I show you in this PR: https://github.com/lakehq/sail/pull/715
Build error: https://github.com/lakehq/sail/actions/runs/16804418669/job/47593139862?pr=715

**Comment by lonless9** [comment]
> Sorry, I didn't explain myself well 😅 What I wanted to say is that there is an error when the GitHub action runs clippy that which has nothing to do with any of the code I did.

Sorry for the inconvenience! I'll fix this.

**Comment by linhr** [comment]
This is probably fixed in #720.

This happens since we use the latest Rust version in GitHub Actions, so whenever a new version is available, there may be new lints that cause our build to fail.

**Review by lonless9 — APPROVED** [review]
This is great! Thanks for your contribution!
